### PR TITLE
KAFKA-10574: Fix infinite loop in Values::parseString

### PR DESCRIPTION
**Pull request by C0urante**
[Jira](https://issues.apache.org/jira/browse/KAFKA-10574)

The special byte sequence `0xEF, 0xBF, 0xBF`, when parsed as a UTF-8 string, causes the `StringCharacterIterator` to return `CharacterIterator.DONE` from its `next()`, `current()`, and other similar methods. This caused an infinite loop in the `Values` class whenever that byte sequence was encountered.

The fix is pretty simple. To see if we're at the end of a string, we compare `StandardCharacterIterator::getIndex` to `StandardCharacterIterator::getEndIndex`, instead of comparing the last-read character to `CharacterIterator.DONE` (since that character may occur in strings that we're parsing).

I've added a unit test that replicates this bug (when the fix in the `Values` class is not present). I gave it a timeout of five seconds in case someone accidentally re-creates the infinite loop in order to save some time and potential confusion.

All existing unit tests for the `Values` class pass with this change.
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Review by gharris1727 — APPROVED** [review]
LGTM!

**Comment by C0urante** [comment]
@rhauch @kkonstantine can one of you take a look?

File: connect/api/src/test/java/org/apache/kafka/connect/data/ValuesTest.java
Comment (kkonstantine) [inline connect/api/src/test/java/org/apache/kafka/connect/data/ValuesTest.java]: we need a comment here to explain things.

**Review by kkonstantine — APPROVED** [review]
LGTM. The fixes makes sense to me but I'm not intimately familiar with this class. 
@rhauch, any concerns with the new check?


@C0urante I'd like to ask for a comment to the test so it's not as cryptic as assuming what these bytes mean for the given encoding.

File: connect/api/src/test/java/org/apache/kafka/connect/data/ValuesTest.java
Comment (C0urante) [reply connect/api/src/test/java/org/apache/kafka/connect/data/ValuesTest.java]: 👍   good call, will add

**Review by kkonstantine — COMMENTED** [review]
fyi: jobs fail with checkstyle errors

**Review by kkonstantine — CHANGES_REQUESTED** [review]
Need to fix checkstyle before merging.

**Comment by C0urante** [comment]
Thanks @kkonstantine, I've added a comment and addressed the Checkstyle issues.

**Review by rhauch — APPROVED** [review]
LGTM, pending a green build. Thanks, @C0urante.

**Review by kkonstantine — APPROVED** [review]
LGTM after the latest fixes

**Comment by C0urante** [comment]
Hi @rhauch @kkonstantine, this PR has been approved, all merge conflicts have been resolved, and every build is green. Would it be possible to merge sometime soon so that it can be backported to bug fix releases and avoid accruing merge conflicts in case the files touched here are modified in other PRs? Thanks!

**Comment by sfc-gh-japatel** [comment]
Hello, I am trying to repro this issue in a connect environment, any guidance on what could be the message to hit this in 2.3.1 and eventually an infinite loop?

**Pull request merged**
